dpe/src/x509.rs

```diff
@@ -483,6 +563,136 @@ impl X509CertWriter<'_> {
         Ok(bytes_written)
     }
 
+    fn encode_fwid(&mut self, tci: &TciMeasurement) -> Result<usize, DpeErrorCode> {
+        let mut bytes_written = self.encode_byte(Self::SEQUENCE_TAG)?;
+        bytes_written +=
+            self.encode_size_field(Self::get_fwid_size(&tci.0, /*tagged=*/ false)?)?;
+
+        // hashAlg OID
+        bytes_written += self.encode_byte(Self::OID_TAG)?;
+        bytes_written += self.encode_size_field(Self::HASH_OID.len())?;
+        bytes_written += self.encode_bytes(Self::HASH_OID)?;
+
+        // digest OCTET STRING
+        bytes_written += self.encode_byte(Self::OCTET_STRING_TAG)?;
+        bytes_written += self.encode_size_field(tci.0.len())?;
+        bytes_written += self.encode_bytes(&tci.0)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a tcg-dice-TcbInfo structure
+    ///
+    /// https://trustedcomputinggroup.org/wp-content/uploads/TCG_DICE_Attestation_Architecture_r22_02dec2020.pdf
+    ///
+    /// TcbInfo makes use of implicitly encoded types. This means the tag
+    /// denotes that the type is implicit (8th bit set) and number of the
+    /// field. For example, "Implicit tag number 2" would be encoded with
+    /// the tag 0x82 for primitive types.
+    ///
+    /// For constructed types (SEQUENCE, SEQUENCE OF, SET, SET OF) the 6th
+    /// bit is also set. For example, "Implicit tag number 2" would be encoded
+    /// with tag 0xA2 for constructed types.
+    fn encode_tcb_info(&mut self, node: &TciNodeData) -> Result<usize, DpeErrorCode> {
+        let tcb_info_size = Self::get_tcb_info_size(node, /*tagged=*/ false)?;
+        // TcbInfo sequence
+        let mut bytes_written = self.encode_byte(Self::SEQUENCE_TAG)?;
+        bytes_written += self.encode_size_field(tcb_info_size)?;
+
+        // fwids SEQUENCE OF
+        // IMPLICIT [6] Constructed
+        let fwid_size = Self::get_fwid_size(&node.tci_current.0, /*tagged=*/ true)?;
+        bytes_written += self.encode_byte(0xA6)?;
```

Comment:
Maybe it would be more self-documenting if they have `const` describing the upper bits OR'd together.
```suggestion
        bytes_written += self.encode_byte(IMPLICIT | CONSTRUCTED | 0x06)?;
```

There are probably better names for them.